Fix/auth backend

rhenanbartels opened the pull request:
Este PR está relacionado a issu #511  e implementa:
1. a validação do username no momento do Cadastro;
      - A partir de agora apenas **letras**, **número** e **_** são permitidos. Caso contrário um erro de validação de Form é levantado.
2. descrição dos erros do Form de login no template HTML.
3. atualização do script de migração de usernames com '@'.
   -  adiciona testes para as funções;
   - atualiza os registro no banco;
   - escreve um arquivo CSV no path '/data/fixed-usernames.csv' com os registros atualizados.



Para rodar o comando de atualização dos usernames:

```
python manage.py runscript migrate_wrong_usernames
```

turicas committed:
Adiciona função que migra usernames (parcialmente)

turicas committed:
Altera maneira de buscar usuários/emails

rhenanbartels committed:
Valida caracteres especiais em username durante SignUp

rhenanbartels committed:
Mostra mensagens de erro do Form

rhenanbartels committed:
Refatora e implementa testes para script de correção de usernames

rhenanbartels committed:
Transforma script em RunsScript

rhenanbartels commented:
Fala @marcmatias,

pode dar uma olhada nesse commit? 

1f26787e3803a8b930481e2bccfd3da40c68ec

A ideia foi colocar um @ no início do input de username, tanto no login, quanto no signup.

![image](https://user-images.githubusercontent.com/2186646/113066918-c6693080-9191-11eb-824f-6b46bf322944.png)

turicas commented on brasilio_auth/forms.py:
```diff
@@ -24,7 +26,13 @@ class Meta:
     def clean_username(self):
         username = self.cleaned_data.get("username", "")
-        return username.lower()
+        username = username.lower()
```
Sei que esse é o comportamento de antes, mas acho que não deveríamos mudar o username aqui para minúsculas. A única mudança seria `username = username.strip()`. Dessa forma, a busca na expressão regular ficaria: `re.search(..., username.lower())` ou então teria que adicionar "A-Z" na expressão regular.

rhenanbartels replied:
solved by be44d62d7fcb99ae04b2ff04

turicas commented on brasilio_auth/forms.py:
```diff
@@ -24,7 +26,13 @@ class Meta:
         username = self.cleaned_data.get("username", "")
-        return username.lower()
+        username = username.lower()
+        non_valid_chars = re.search(r"[^a-z0-9_]", username)
```
Ao usar expressões regulares em funções que serão executadas repetidas vezes, o ideal é compilar a expressão no início do comprar e usar a versão compilada. Nesse caso, a execução de uma regexp compilada demora apenas 24% do tempo de uma não compilada:

```
In [1]: import re

In [2]: %timeit re.search(r"[^a-z0-9_]", "turicas")
606 ns ± 30.5 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [3]: REGEXP_USERNAME = re.compile(r"[^a-z0-9_]")

In [4]: %timeit REGEXP_USERNAME.search("turicas")
145 ns ± 9.22 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [5]: 145/606
Out[5]: 0.23927392739273928
```

rhenanbartels replied:
solved by 985cc1d3bae5294d56f2

turicas commented on brasilio_auth/templates/brasilio_auth/login.html:
```diff
@@ -7,9 +7,18 @@
-        </p>
+          {% for field in form %}
+              {% for error in field.errors %}
+                 <p class="red-text text-lighten-3">
+                      {{ error|escape }}
+                  </p>
+              {% endfor %}
+          {% endfor %}
+          {% for error in form.non_field_errors %}
+             <p class="red-text text-lighten-3">
+                 {{ error|escape }}
+              </p>
```
Faltou seguir a indentação aqui.

rhenanbartels replied:
solved by 20f53519b1f520

turicas commented on brasilio_auth/scripts/migrate_wrong_usernames.py:
```diff
@@ -38,23 +39,29 @@ def possible_usernames(username, email):
+            possible = possible_usernames(user.username, user.email)
+            changed = False
+            for username in possible:
+                if not User.objects.filter(username=username).exists() and not changed:
```
Aqui não precisaria do `not changed` se tivesse um `break` no fim do bloco do `if`, depois do `user.save()`.

rhenanbartels replied:
solved by 8d6bddccd0f6b9fe32

turicas commented on brasilio_auth/tests/test_scripts.py:
```diff
@@ -0,0 +1,124 @@
+        assert possible_values == expected
+
+
+class TestReplaceUsernameWithSuggestions(TestCase):
```
Fiacaram muito bons os testes! o/

turicas reviewed (changes requested):
Muito bom! Coloquei alguns comentários de alteração, mas são relativamente simples, apenas com intuito de melhorar a legibilidade e tempo de execução.

rhenanbartels committed:
Ajusta identação em html

rhenanbartels committed:
Utilza break ao invés de changed

rhenanbartels committed:
Utiliza versão compilada de regex

rhenanbartels committed:
Remove trecho que força username para caixa baixa

marcmatias commented:
> Fala @marcmatias,
> 
> pode dar uma olhada nesse commit?
> 
> [1f26787](https://github.com/turicas/brasil.io/commit/1f26787e3803a8b930481e2bccfd3da40c68ec27)
> 
> A ideia foi colocar um @ no início do input de username, tanto no login, quanto no signup.
> 
> ![image](https://user-images.githubusercontent.com/2186646/113066918-c6693080-9191-11eb-824f-6b46bf322944.png)

Opa, tranquilo. Um modo que exigiria menos código seria pondo o @ como bg no inputfield. subi aqui: [Codepen exemplo](https://codepen.io/marcmatias/pen/RwKoBwY) resultado semelhante ao apresentado.

rhenanbartels committed:
Utiliza background para colocar @ em input de username

turicas committed:
Corrige todos os usernames inválidos

turicas committed:
Não permite usuários somente com dígitos e pontuação

Isso evita que usuários coloquem seus CPFs como usernames (não queremos
armazenar dados sensíveis).

turicas committed:
Garante username único com novo teste

turicas committed:
Corrige testes (faltava a API em ALLOWED_HOSTS)

turicas committed:
Remove forms desnecessários

rhenanbartels committed:
Roda todos os comandos pelo docker-compose

rhenanbartels committed:
Ignore erro do flake8
